### PR TITLE
Screentip images

### PR DESCRIPTION
**Pull request by SandPoot**
## About The Pull Request
Adds some nice little mouse button indicators (optional), this is a port of my version at [Citadel-Station-13/15973](https://github.com/Citadel-Station-13/Citadel-Station-13/pull/15973).

Images:
Image allowed
dreamseeker_Uhoqrh7GQE
dreamseeker_JOWbi76IFw
image
Image disallowed
image
image
image
Preferences
image

**There is no more intent coloring**

## Why It's Good For The Game
Quick identification of what to do. Eh, i dunno if it's getting accepted, mothblocks asked me to port it over.
## Changelog
:cl:
image: Changed screentips icons to something a lot fancier unified with the LMB / RMB text.
qol: There are accessibility preferences to disable this! Look for "Screentip context with images" beside the other screentip preferences.
/:cl:

### CODE REVIEW

**Comment by SandPoot** [comment]
Review auto-requested from every single user, how to get gitbanned speedrun.

**Review by Mothblocks — CHANGES_REQUESTED** [review]
I really do not like how every single consumer of screentips needs to care about this when only a small handful of interactions actually care about "intent", that's the sign of an API that can be improved. On that note, "intent" does not exist on tg and thus pretending it is in screentips is a mis-step.

I'm not sure this needs the colors at all--it's not obvious to me as a player who played before intents why right-clicking on a human is blue while ejecting a beaker is grey. Both are just actions you do with right-click.

I don't think this needs more than the base images.

**Comment by SandPoot** [comment]
I'll change it when i'm free, a bunch of changes will have to be undone.

**Review by Wallemations — APPROVED** [review]
Cool shit but what the fuck is that mouse cursor

**Comment by Wallemations** [comment]
Also, I for one think that the mouse buttons being the same color as the original intents when used on a living thing is a cute callback to olden times

I don't think that players will overthink it and wonder why it's suddenly a different color, if anything they might just see that as an indicator that the thing you're mousing over is a mob rather than any sort of deeper meaning.

**Comment by SandPoot** [comment]
> Cool shit but what the fuck is that mouse cursor

https://www.deviantart.com/gsergiu/art/Sharp-Cursors-green-312285380

**Comment by SandPoot** [comment]
Also that reminds me, the whole original concept of all this is that i loved the wiki indicators for things,
image
so me and a friend made a couple sprites that had a couple iterations until we settled for those final ones

**Comment by EricZilla** [comment]
Colorwise, grey for normal mode and red for combat intent (fight me its still an intent) would probably be all we need

**Comment by CoiledLamb** [comment]
> Also that reminds me, the whole original concept of all this is that i loved the wiki indicators for things, image so me and a friend made a couple sprites that had a couple iterations until we settled for those final ones

that's so cute!

**Comment by MrDoomBringer** [comment]
good change good for accessability makes it easier to see at a glance looks beautiful but wtffff who uses a custom cursor lmao

**Comment by SandPoot** [comment]
Almost done with requested changes

**Comment by SandPoot** [comment]
> I really do not like how every single consumer of screentips needs to care about this when only a small handful of interactions actually care about "intent", that's the sign of an API that can be improved. On that note, "intent" does not exist on tg and thus pretending it is in screentips is a mis-step.
> 
> I'm not sure this needs the colors at all--it's not obvious to me as a player who played before intents why right-clicking on a human is blue while ejecting a beaker is grey. Both are just actions you do with right-click.
> 
> I don't think this needs more than the base images.

Changes done.

File: code/__HELPERS/screentips.dm [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#define HINT_ICON_FILE 'icons/ui_icons/screentips/cursor_hints.dmi'
+
+// Generate intent icons
+GLOBAL_LIST_INIT_TYPED(screentip_context_icons, /image, prepare_screentip_context_icons())
+
+/proc/prepare_screentip_context_icons()
+	. = list()
+	for(var/state in icon_states(HINT_ICON_FILE))
+		.[state] = image(HINT_ICON_FILE, icon_state = state)
```
Comment (Mothblocks) [inline code/__HELPERS/screentips.dm:9]: Avoid trivial uses of `.`, this can just be `output`

File: code/__HELPERS/screentips.dm [hunk 2]
```diff
@@ -0,0 +1,33 @@
+#define HINT_ICON_FILE 'icons/ui_icons/screentips/cursor_hints.dmi'
+
+// Generate intent icons
```
Comment (Mothblocks) [inline code/__HELPERS/screentips.dm:3]: Change this instead to `/// A description of what this list is`, including the information that its a mapping of icon state to icon (helpfully even what those states are?).

File: code/__HELPERS/screentips.dm [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ * Args:
+ * - context = list (REQUIRED)
+ * 	- Must contain key
+ * - key = string (REQUIRED)
+ * - allow_image = boolean (not required)
+*/
+/proc/build_context(list/context, key, allow_image)
+	if(!(length(context) && context[key] && key))
+		return ""
+	// Get everything but the mouse button, may be empty
```
Comment (Mothblocks) [inline code/__HELPERS/screentips.dm:22]: These comments should give a consistent example, like "e.g. Alt LMB takes `Alt `"

File: code/_onclick/hud/hud.dm [hunk 4]
```diff
@@ -71,6 +71,8 @@ GLOBAL_LIST_INIT(available_ui_styles, list(
 	/// had with a proc call, especially on one of the hottest procs in the
 	/// game (MouseEntered).
 	var/screentips_enabled = SCREENTIP_PREFERENCE_ENABLED
+	/// Whether to use text or images for click hints
```
Comment (Mothblocks) [inline code/_onclick/hud/hud.dm:74]:
```suggestion
	/// Whether to use text or images for click hints.
	/// Same behavior as `screentips_enabled`--very hot, updated when the preference is updated.
```

File: code/game/atoms.dm [hunk 5]
```diff
@@ -2056,49 +2056,51 @@
 				| (held_item && SEND_SIGNAL(held_item, COMSIG_ITEM_REQUESTING_CONTEXT_FOR_TARGET, context, src, user))
 
 			if (contextual_screentip_returns & CONTEXTUAL_SCREENTIP_SET)
+				var/screentip_images = active_hud.screentip_images
 				// LMB and RMB on one line...
-				var/lmb_text = (SCREENTIP_CONTEXT_LMB in context) ? "[SCREENTIP_CONTEXT_LMB]: [context[SCREENTIP_CONTEXT_LMB]]" : ""
-				var/rmb_text = (SCREENTIP_CONTEXT_RMB in context) ? "[SCREENTIP_CONTEXT_RMB]: [context[SCREENTIP_CONTEXT_RMB]]" : ""
+				var/lmb_text = build_context(context, SCREENTIP_CONTEXT_LMB, screentip_images)
```
Comment (Mothblocks) [inline code/game/atoms.dm:2061]: Let's test merge this for a bit because this proc is flaming hot and I am worried about the effects of so many proc calls being added to it, CC @LemonInTheDark

Comment (Mothblocks) [reply code/game/atoms.dm:2061]: Like it is very possible that we just stick to monochrome images to make this simpler even

File: code/__HELPERS/screentips.dm
Comment (Mothblocks) [inline code/__HELPERS/screentips.dm:1]: Nit: This file should be somewhere in `code/datums/screentips`

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Comment by Mothblocks** [comment]
TM'ing on sybil

**Comment by Mothblocks** [comment]
Lemon is out of commission right now so I'm going to wait for them because they're a lot better at analyzing profiles than me, sorry for delay

File: code/game/atoms.dm (hunk 5 above)
Comment (LemonInTheDark) [reply code/game/atoms.dm:2080]: this is not giving me great vibes I'll admit

Comment (LemonInTheDark) [reply code/game/atoms.dm:2080]: will check in detail tomorrow/day after

**Comment by LemonInTheDark** [comment]
will go over later, vibes funny, smells like something that should be baked in

**Comment by LemonInTheDark** [comment]
here's cost over the last few rounds (new stuff is around about 210700)
the added cost appears to be minimal, likely because the case isn't that common. about 2 seconds each round right now. honestly if we needed to optimize this proc there's other things we could do.
it'll likely be more of an issue as we add more usages, but this isn't poop dimension level so it's not a massive headache

image

image

TL;DR: minimal impact, worth keeping an eye on, will likely worsen as we add more common uses of screentips but it can't be that bad given what's already attached to the system

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
minor complaint
